src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,321 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar, Union
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = Union[list, np.ndarray]
+
+
+class RingBuffer(Generic[T]):
+    """A ring buffer with a fixed size.
+
+    Should work with most backends, tested with list and np.ndarrays.
+    """
+
+    class NoElements(Exception):
+        """Exception class, thrown when no elements are available when calling pop()."""
+
+    def __init__(self, container: Container) -> None:
+        """Initialize the ring buffer with the given container.
+
+        Args:
+            container: Container to store the data in.
+        """
+        self._container = container
+        self._write_index = 0
+        self._read_index = 0
+        self._len = 0
+
+    def __len__(self) -> int:
+        """Get current amount of elements.
+
+        Returns:
+            the amount of items that this container currently holds.
```

Comment:
Silly but try to always start these with caps for more consistency: "The amount of ...".